enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/v3_3/runtime/interpreted/EnterprisePipeBuilder.scala

```diff
@@ -231,12 +231,23 @@ class EnterprisePipeBuilder(fallback: PipeBuilder,
     implicit val table: SemanticTable = context.semanticTable
 
     val id = idMap.getOrElse(plan, new Id)
+    val pipeline = pipelines(plan)
 
     plan match {
       case Apply(_, _) =>
         ApplyRegisterPipe(lhs, rhs)(id)
 
+      case _: SemiApply |
+           _: AntiSemiApply =>
+        fallback.build(plan, lhs, rhs)
+
+      case _: CartesianProduct =>
+        val lhsPlan = plan.lhs.get
+        val lhsLongCount = pipelines(lhsPlan).numberOfLongs
+        val lhsRefCount = pipelines(lhsPlan).numberOfReferences
+        CartesianProductRegisterPipe(lhs, rhs, lhsLongCount, lhsRefCount, pipeline)(id)
```

Comment:
The lhsLongCount and lhsRefCount would be known by the lhsCtx inside internalCreateResults, so why do we need to pass it in here?

Reply:
I agree it's ugly, but even though in the pipe we get is a `PrimitiveExecutionContext` it is passed in as the base `ExecutionContext`, so we'd need to do something like 
```
rhsCtx.copyTo(context, lhsCtx.asInstanceOf[PrimitiveExecutionContext].pipeline.numberOfLongs, lhsCtx.asInstanceOf[PrimitiveExecutionContext].pipeline.numberOfReferences)
```